### PR TITLE
feat: scheduler (7/): add initial parts of scheduler framework

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR is part of the PRs that implement the Fleet workload scheduling.

It features the scaffolding for scheduler framework.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

N/A.

To accommodate the uncertainties forward, and considering the fact that there is no easy way to fake a controller runtime manager, this PR does not include unit test (sanity check) for NewFramework(). Tests will be checked in with future PRs.

### CODE REVIEW

File: pkg/scheduler/framework/framework.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	fleetv1 "go.goms.io/fleet/apis/v1"
+	"go.goms.io/fleet/pkg/scheduler/framework/parallelizer"
+)
+
+const (
+	// eventRecorderNameTemplate is the template used to format event recorder name for a scheduler framework.
+	eventRecorderNameTemplate = "scheduler-framework-%s"
+)
+
+// Handle is an interface which allows plugins to set themselves up with the scheduler framework.
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/framework.go:27]:
```suggestion
// Handle is an interface which provides data and some tools that plugins can use. The plugins will set themselves up with the scheduler framework.
```
is my understanding correct？ PTAL

**Review by zhiying-lin — COMMENTED** [review]
minor comments, others LGTM :)

File: pkg/scheduler/framework/framework.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	fleetv1 "go.goms.io/fleet/apis/v1"
+	"go.goms.io/fleet/pkg/scheduler/framework/parallelizer"
+)
+
+const (
+	// eventRecorderNameTemplate is the template used to format event recorder name for a scheduler framework.
+	eventRecorderNameTemplate = "scheduler-framework-%s"
```
Comment (ryanzhang-oss) [inline pkg/scheduler/framework/framework.go:24]: is this event msg readable for user?

File: pkg/scheduler/framework/framework.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+// Handle is an interface which allows plugins to set themselves up with the scheduler framework.
+type Handle interface {
+	// Client returns a cached client.
+	Client() client.Client
+	// Manager returns a controller manager; this is mostly used for setting up a new informer
+	// (indirectly) via a reconciler.
+	Manager() ctrl.Manager
+	// APIReader returns an uncached read-only client, which allows direct (uncached) access to the API server.
+	APIReader() client.Reader
```
[the remark this answers is not in the document]
Comment (ryanzhang-oss) [reply pkg/scheduler/framework/framework.go:36]: directReader or unCachedReader? I am not sure "APIReader" convey the meaning.

File: pkg/scheduler/framework/framework.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+)
+
+const (
+	// eventRecorderNameTemplate is the template used to format event recorder name for a scheduler framework.
+	eventRecorderNameTemplate = "scheduler-framework-%s"
+)
+
+// Handle is an interface which allows plugins to access some shared structs (e.g., client, manager)
+// and set themselves up with the scheduler framework (e.g., sign up for an informer).
+type Handle interface {
```
Comment (ryanzhang-oss) [inline pkg/scheduler/framework/framework.go:29]: In general, I think interface name is better to be a noun while the function name is a verb. ie

type Writer interface {
    Write(p []byte) (n int, err error)
}



We have the exact opposite pattern here.

File: pkg/scheduler/framework/framework.go (hunk 2 above)
Comment (michaelawyu) [reply pkg/scheduler/framework/framework.go:24]: Hi Ryan! Ah, this isn't the event message, but the name of the event recorder in use by the scheduler framework.

File: pkg/scheduler/framework/framework.go (hunk 4 above)
Comment (michaelawyu) [reply pkg/scheduler/framework/framework.go:29]: The name is also from Kubernetes; `Handler` would be a little bit weird, it's more of a handle that allows plugins to access the framework (in a limited manner).

File: pkg/scheduler/framework/framework.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/scheduler/framework/framework.go:36]: I will use uncachedReader then.

File: pkg/scheduler/framework/framework.go [hunk 5]
```diff
@@ -133,8 +134,8 @@ func (f *framework) Manager() ctrl.Manager {
 }
 
 // APIReader returns the (uncached) read-only client in use by the scheduler framework.
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/framework.go:136]: update the comment?

Comment (michaelawyu) [reply pkg/scheduler/framework/framework.go]: Done.

**Review by zhiying-lin — APPROVED** [review]

**Comment by michaelawyu** [comment]
Hi Ryan! I took the liberty to merge this PR first; if you have any further concerns about it, please let me know.

**Pull request merged**